src/core/api/clock.ts

```diff
@@ -198,102 +200,141 @@ function getMediaInfos(
 }
 
 /**
- * Infer stalled status of the media based on:
- *   - the return of the function getMediaInfos
- *   - the previous timings object.
- *
- * @param {Object} prevTimings - Previous timings object. See function to know
- * the different properties needed.
- * @param {Object} currentTimings - Current timings object. This does not need
- * to have every single infos, see function to know which properties are needed.
- * @param {Object} options
- * @returns {Object|null}
+ * @returns {function}
  */
-function getStalledStatus(
+function getStallDetector(): (
   prevTimings : IClockTick,
   currentTimings : IMediaInfos,
   { withMediaSource, lowLatencyMode } : IClockOptions
-) : IStalledStatus {
-  const { state: currentState,
-          currentTime,
-          bufferGap,
-          currentRange,
-          duration,
-          paused,
-          readyState,
-          ended } = currentTimings;
-
-  const { stalled: prevStalled,
-          state: prevState,
-          currentTime: prevTime } = prevTimings;
-
-  const fullyLoaded = hasLoadedUntilTheEnd(currentRange, duration, lowLatencyMode);
-
-  const canStall = (readyState >= 1 &&
-                    currentState !== "loadedmetadata" &&
-                    prevStalled === null &&
-                    !(fullyLoaded || ended));
-
-  let shouldStall : boolean | undefined;
-  let shouldUnstall : boolean | undefined;
-
-  if (withMediaSource) {
-    if (canStall &&
-        (bufferGap <= (lowLatencyMode ? STALL_GAP.LOW_LATENCY : STALL_GAP.DEFAULT) ||
-         bufferGap === Infinity || readyState === 1)
-    ) {
-      shouldStall = true;
-    } else if (prevStalled !== null &&
-               readyState > 1 &&
-               ((bufferGap < Infinity &&
-                 bufferGap > getResumeGap(prevStalled, lowLatencyMode)) ||
-                fullyLoaded || ended)
-    ) {
-      shouldUnstall = true;
-    }
-  }
+) => IStalledStatus {
+  let isFreezingSince: undefined|number;
```

Comment:
I don't get why `isFreezingSince` is needed. Isn't the timestamp added to `stalled` sufficient?

Also, I find that the code in the clock is becoming very hard to understand.